### PR TITLE
LGA-196: Log to stdout

### PR DESCRIPTION
**Pull request by rowlando**
## What does this pull request do?

This pull request configures "nginx" and "uwsgi" to log to stdout in JSON format.

## Any other changes that would benefit highlighting?

Note that "nginx" does not log the timestamp to millisecond resolution. Instead of `$time_iso8601`, we could choose to log `$msec`, which is defined as:

>time in seconds with a milliseconds resolution at the time of the log write 

Below is an example of the log entries for nginx, uwsgi and the application. 

### Nginx log entry

```
{
  "@timestamp": "2018-09-19T11:33:01+01:00",
  "@fields": {
    "process_name": "nginx",
    "pid": 27,
    "remote_addr": "192.168.144.1",
    "remote_user": "-",
    "body_bytes_sent": 17,
    "request_time": 0.003,
    "request_uri": "/session_keep_alive",
    "status": 200,
    "request": "GET /session_keep_alive HTTP/1.1",
    "request_method": "GET",
    "http_referrer": "http://localhost:5000/session-expired",
    "http_user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36",
    "bytes_sent": 453,
    "gzip_ratio": "-",
    "http_host": "localhost",
    "sent_http_location": "-",
    "server_name": "~^(.+)$",
    "server_port": "80",
    "upstream_addr": "unix:////tmp/public.sock",
    "upstream_response_length": "17",
    "upstream_response_time": "0.010",
    "upstream_status": "200"
  }
}
```

### uwsgi log entry

```
{
  "process_name": "uwsgi",
  "timestamp_msec": 1537353181294496,
  "method": "GET",
  "uri": "/session_keep_alive",
  "proto": "HTTP/1.1",
  "status": 200,
  "referer": "http://localhost:5000/session-expired",
  "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36",
  "remote_addr": "192.168.144.1",
  "http_host": "localhost:5000",
  "pid": 35,
  "worker_id": 1,
  "core": 0,
  "async_switches": 1,
  "io_errors": 0,
  "rq_size": 0,
  "rs_time_ms": 2,
  "rs_size": 288,
  "rs_header_size": 271,
  "rs_header_count": 3
}
```

### Application log entry

```
{
  "@fields": {
    "exception": [
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 1817, in wsgi_app\n    response = self.full_dispatch_request()\n",
      "  File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 1477, in full_dispatch_request\n    rv = self.handle_user_exception(e)\n",
      "  File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 1381, in handle_user_exception\n    reraise(exc_type, exc_value, tb)\n",
      "  File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 1475, in full_dispatch_request\n    rv = self.dispatch_request()\n",
      "  File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 1461, in dispatch_request\n    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"./cla_public/apps/base/views.py\", line 188, in get_started\n    logger.info({\"account\": 123, \"ip\": \"172.20.19.18\"})\n",
      "NameError: global name 'logger' is not defined\n"
    ],
    "threadName": "uWSGIWorker2Core0",
    "name": "cla_public.app",
    "thread": 139654685837056,
    "relativeCreated": 29431.34093284607,
    "process": 37,
    "args": [],
    "module": "app",
    "funcName": "log_exception",
    "levelno": 40,
    "processName": "MainProcess",
    "created": 1537349043.551323,
    "msecs": 551.3229370117188,
    "pathname": "/usr/local/lib/python2.7/dist-packages/flask/app.py",
    "filename": "app.py",
    "levelname": "ERROR",
    "lineno": 1423
  },
  "@timestamp": "2018-09-19T09:24:03.551628Z",
  "@source_host": "d63c5ea1d9ba",
  "@message": "Exception on /start [GET]"
}
```

### CODE REVIEW

File: docker/cla_public.ini [hunk 1]
```diff
@@ -11,8 +11,6 @@ processes = 2
 chdir = /home/app/flask
 module = cla_public.wsgi
 callable = app
-req-logger = file:/var/log/wsgi/request.log
-logger = file:/var/log/wsgi/error.log
```
Comment (sldblog) [inline docker/cla_public.ini]: Would specifying `file:/dev/stdout` work?

Comment (rowlando) [reply docker/cla_public.ini]: No, that doesn't seem to work. If those lines are removed, uwsgi logs to stdout okay. Here you can see 2 logs entries, the first one is from uwsgi and the second from nginx.

```
localhost:5000 [pid: 36|app: 0|req: 10/14] 192.168.144.1 () {52 vars in 1356 bytes} [Tue Sep 18 15:00:52 2018] GET /scope/diagnosis/ => generated 20624 bytes in 144 msecs (HTTP/1.1 200) 5 headers in 500 bytes (1 switches on core 0)
{ "@timestamp": "2018-09-18T15:00:52+01:00", "@fields": { "remote_addr": "192.168.144.1", "remote_user": "-", "body_bytes_sent": 6503, "request_time": 0.145, "request_uri": "/scope/diagnosis/", "status": 200, "request": "GET /scope/diagnosis/ HTTP/1.1", "request_method": "GET", "http_referrer": "-", "http_user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36", "bytes_sent": 7197, "gzip_ratio": "3.18", "http_host": "localhost", "sent_http_location": "-", "server_name": "~^(.+)$", "server_port": "80", "upstream_addr": "unix:////tmp/public.sock", "upstream_response_length": "20624", "upstream_response_time": "0.140", "upstream_status": "200" } }
```

I need to customise the logs as per the ticket JSON.

Comment (sldblog) [reply docker/cla_public.ini]: 👍 

:badpokerface: In #755, I ended up converting the `wsgi` logs to logstash format as part of simplifying the configuration -- so rebasing might change the behaviour. Sorry :(

File: docker/cla_public.ini [hunk 2]
```diff
@@ -11,8 +11,9 @@ processes = 2
...
+logger-req=stdio
+log-req-encoder=json {"process_name": "uwsgi", "timestamp": "${strftime:%%Y-%%m-%%dT%%H:%%M:%%S}", ${msg}}
```
Comment (sldblog) [inline docker/cla_public.ini]: 👏 TIL

💭 Would it be possible to annotate this with milliseconds as well?

Comment (sldblog) [reply docker/cla_public.ini]: According to the documentation at https://uwsgi-docs.readthedocs.io/en/latest/LogEncoders.html#core-encoders we could use `${micros}`

> `${micros}` the current unix time in microseconds

E.g. `json {"process_name": "uwsgi", "timestamp_msec": ${micros}, "message": "${msg}"}`

What do you think?

Comment (rowlando) [reply docker/cla_public.ini]: Thanks for the tip!

File: docker/nginx.conf [hunk 3]
```diff
@@ -35,6 +35,8 @@ http {
 
         log_format logstash_json '{ "@timestamp": "$time_iso8601", '
                                  '"@fields": { '
+                                 '"process_name": "nginx", '
+                                 '"pid": $pid, '
```
Comment (sldblog) [inline docker/nginx.conf]: 👏

File: docker/cla_public.ini [hunk 4]
```diff
@@ -11,8 +11,9 @@ processes = 2
 chdir = /home/app/flask
 module = cla_public.wsgi
 callable = app
-req-logger = file:/var/log/wsgi/request.log
-logger = file:/var/log/wsgi/error.log
+logger-req=stdio
+log-req-encoder=json {"process_name": "uwsgi", "timestamp_msec": ${micros}, ${msg}}
```
Comment (sldblog) [inline docker/cla_public.ini]: 👏

**Pull request merged**